numcodecs/compat.py

```diff
@@ -160,6 +167,12 @@ def ensure_bytes(buf):
 def ensure_text(s, encoding='utf-8'):
     if not isinstance(s, text_type):
         s = ensure_contiguous_ndarray(s)
+
+        # Force CuPy arrays to NumPy arrays
+        # as they support the buffer protocol
+        if isinstance(s, cp.ndarray):
+            s = s.get()
```

Comment:
One way to solve this might be to call `ensure_bytes` here. That would guarantee we have something we can perform text decoding on. It would also generalize for our use case here.

On the flip side that does incur a copy. Though this may be pretty inexpensive in the grand scheme of things.

Reply:
I think in Dask we do a check like `if hasattr(s, "get"): s = s.get()`.  This is a little non-specific, but it works well in practice and doesn't require the cupy import.

Reply:
Just for context, this is what the `ensure_bytes` change would look like ( https://github.com/zarr-developers/numcodecs/pull/213 ).

Reply:
It's also possible we can get away with dropping this part entirely. This function is mainly used in conjunction with reading things (like JSON). So ensuring it can work with CuPy is probably unnecessary.

Reply:
Yes not sure you need to worry about cupy arrays here as this is mainly for JSON reading IIRC.